packages/vite/src/node/build.ts

```diff
@@ -997,22 +1020,33 @@ function isExternal(id: string, test: string | RegExp) {
   }
 }
 
-function injectSsrFlagToHooks(plugin: Plugin): Plugin {
+function injectEnvironmentToHooks(
+  plugin: Plugin,
+  environment?: BuildEnvironment,
+): Plugin {
   const { resolveId, load, transform } = plugin
   return {
     ...plugin,
-    resolveId: wrapSsrResolveId(resolveId),
-    load: wrapSsrLoad(load),
-    transform: wrapSsrTransform(transform),
+    resolveId: wrapEnvironmentResolveId(resolveId, environment),
+    load: wrapEnvironmentLoad(load, environment),
+    transform: wrapEnvironmentTransform(transform, environment),
   }
 }
 
-function wrapSsrResolveId(hook?: Plugin['resolveId']): Plugin['resolveId'] {
+function wrapEnvironmentResolveId(
+  hook?: Plugin['resolveId'],
+  environment?: BuildEnvironment,
+): Plugin['resolveId'] {
   if (!hook) return
 
   const fn = getHookHandler(hook)
   const handler: Plugin['resolveId'] = function (id, importer, options) {
-    return fn.call(this, id, importer, injectSsrFlag(options))
+    return fn.call(
+      injectEnvironmentInContext(this),
```

Comment:
I also saw `this.environment` is undefined in `resolveId` during build. Probably this one missing too?

```suggestion
      injectEnvironmentInContext(this, environment),
```

Reply:
Thanks! We'll need a lot of work on the e2e tests before we can launch this